File: frontend/src/types/smart.ts
export type Tier = 'trivial' | 'simple' | 'standard' | 'complex'
export type AgentId = 'claude' | 'codex' | 'copilot' | 'antigravity' | 'ollama'
export type QuotaReset = 'none' | '5h' | 'daily' | 'weekly'

export interface RouteTarget { provider_id: string; model: string }
export interface TierPolicy { tier: Tier; targets: RouteTarget[] }
export interface BudgetWindow {
  target_key: string
  window: 'day' | 'week' | 'month'
  max_requests: number
  max_cost_usd: number
}
export interface SmartRoutingConfig {
  enabled: boolean
  mode: 'shadow' | 'active'
  thresholds: Record<string, number>
  tiers: TierPolicy[]
  budgets: BudgetWindow[]
  honor_tier_header: boolean
  sticky_tool_loops: boolean
  sticky_ttl_seconds: number
  skip_cooling_providers: boolean
  respect_capabilities: boolean
}
export interface CliAgent {
  id: AgentId
  name: string
  enabled: boolean
  exe_path: string
  default_model: string
  model_by_tier: Record<string, string>
  alt_model_on_quota: string
  supported_tiers: string[]
  agentic: boolean
  priority: number
  cost_weight: number
  max_concurrency: number
  timeout_s: number
  cooldown_s: number
  quota_reset: QuotaReset
  max_credits: number
  extra_args: string[]
}
export interface DelegationConfig {
  enabled: boolean
  workspace_allowlist: string[]
  tier_order: Record<string, string[]>
  max_parallel_jobs: number
  max_attempts: number
  job_retention: number
}
export interface AgentStatus {
  id: string
  installed: boolean
  exe: string
  version: string
  auth: 'ok' | 'auth_error' | 'unknown'
  state: string
  seconds_left: number
  last_signal: string
  last_excerpt: string
  quota: Record<string, { remaining_fraction?: number; reset_time?: string; name?: string; models?: string[]; context_capped?: boolean }>
  default_model: string
  deny_list_present?: boolean | null
  running: number
  checked_at: string
  error: string
}
export interface TargetHealth {
  state: string
  until?: string
  last_signal?: string
  last_excerpt?: string
  consecutive_failures?: number
  total_ok?: number
  total_failed?: number
  seconds_left?: number
}
export interface BudgetUsage extends BudgetWindow {
  requests: number
  cost_usd: number
  exceeded: boolean
}
export interface Recommendation { code: string; message: string; patch?: { provider_id: string; updates: Record<string, unknown> } }
export interface SmartConfigResponse {
  smart: SmartRoutingConfig
  cli_agents: CliAgent[]
  delegation: DelegationConfig
  agents: AgentStatus[]
  health: Record<string, TargetHealth>
  budgets: Record<string, BudgetUsage>
  recommendations: Recommendation[]
  tiers: string[]
}
export interface Classification {
  tier: Tier
  score: number
  intent: string
  reasons: string[]
  source: string
  signals: Record<string, unknown>
}
export interface RouteDecision extends Classification {
  id: string
  chosen_key: string
  chosen_model: string
  would_key: string
  would_model: string
  mode: string
  rejected: [string, string][]
  decision_ms: number
  is_active: boolean
}
export interface DecisionRow extends RouteDecision {
  timestamp: string
  surface: string
  requested_model: string
  prompt_tokens: number
  outcome: string
  outcome_latency_ms?: number
  outcome_error?: string
}
export interface DecisionsSummary {
  period: string
  count: number
  by_tier: Record<string, number>
  by_target: Record<string, number>
  by_source: Record<string, number>
  outcomes: { ok: number; error: number; pending: number }
  agreement_rate: number
  avg_decision_ms: number
}
export interface Attempt {
  agent_id: string; model: string; started_at: string; finished_at?: string | null
  returncode?: number | null; signal: string; duration_s: number; error: string
}
export type JobStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'timeout' | 'cancelled' | 'quota' | 'auth_error'
export interface Job {
  id: string; created_at: string; started_at?: string | null; finished_at?: string | null
  status: JobStatus; mode: 'task' | 'text'; workspace: string; task_preview: string
  tier: string; score: number; reasons: string[]; skipped: string[][]; agent_id?: string | null
  model: string; attempts: Attempt[]; output_tail: string; files_touched: string[]
  error: string; tokens_in: number; tokens_out: number; cost_usd?: number | null; log_path: string
}
export interface JobRequest {
  task: string; workspace: string; mode: 'task' | 'text'; tier_hint?: Tier
  agent_id?: string; model?: string; timeout_s?: number; dry_run?: boolean
}
export interface SseEvent { event: 'status' | 'line' | 'attempt' | 'ping' | 'done'; [key: string]: unknown }

File: backend/app/models/smart.py
"""
Modelos de la gestión inteligente: routing por complejidad (tiers, presupuestos)
y agentes CLI delegables. Todo con defaults neutros: un providers.json viejo carga
sin cambios y `smart.enabled=False` reproduce el comportamiento previo.
"""
from typing import Literal, Optional

from pydantic import BaseModel, Field

Tier = Literal["trivial", "simple", "standard", "complex"]
TIER_ORDER: tuple[str, ...] = ("trivial", "simple", "standard", "complex")

AgentId = Literal["claude", "codex", "copilot", "antigravity", "ollama"]
AGENT_IDS: tuple[str, ...] = ("claude", "codex", "copilot", "antigravity", "ollama")

TargetState = Literal["available", "cooling", "exhausted", "unavailable"]
QuotaReset = Literal["none", "5h", "daily", "weekly"]


def tier_index(tier: str) -> int:
    return TIER_ORDER.index(tier) if tier in TIER_ORDER else -1


class RouteTarget(BaseModel):
    provider_id: str
    model: str = ""  # "" = active_model del provider

    @property
    def key(self) -> str:
        return f"provider:{self.provider_id}"


class TierPolicy(BaseModel):
    tier: Tier
    targets: list[RouteTarget] = Field(default_factory=list)  # orden = preferencia


class BudgetWindow(BaseModel):
    target_key: str  # "provider:<id>" | "cli:<id>"
    window: Literal["day", "week", "month"] = "day"
    max_requests: int = 0  # 0 = sin límite
    max_cost_usd: float = 0.0


class SmartRoutingConfig(BaseModel):
    enabled: bool = False
    mode: Literal["shadow", "active"] = "shadow"
    thresholds: dict[str, int] = Field(default_factory=lambda: {"simple": 25, "standard": 50, "complex": 75})
    tiers: list[TierPolicy] = Field(default_factory=list)
    budgets: list[BudgetWindow] = Field(default_factory=list)
    honor_tier_header: bool = True
    sticky_tool_loops: bool = True
    sticky_ttl_seconds: int = 1800
    skip_cooling_providers: bool = True
    respect_capabilities: bool = True

    def policy_for(self, tier: str) -> Optional[TierPolicy]:
        return next((p for p in self.tiers if p.tier == tier), None)


class CliAgent(BaseModel):
    id: AgentId
    name: str = ""
    enabled: bool = False
    exe_path: str = ""
    default_model: str = ""
    model_by_tier: dict[str, str] = Field(default_factory=dict)
    alt_model_on_quota: str = ""
    supported_tiers: list[str] = Field(default_factory=lambda: ["simple", "standard", "complex"])
    agentic: bool = True
    priority: int = 50
    cost_weight: float = 1.0
    max_concurrency: int = 1
    timeout_s: int = 600
    cooldown_s: int = 900
    quota_reset: QuotaReset = "none"
    max_credits: int = 0
    extra_args: list[str] = Field(default_factory=list)

    @property
    def key(self) -> str:
        return f"cli:{self.id}"

    def model_for(self, tier: str, requested: str = "") -> str:
        return requested or self.model_by_tier.get(tier, "") or self.default_model


class DelegationConfig(BaseModel):
    enabled: bool = False
    workspace_allowlist: list[str] = Field(default_factory=list)
    tier_order: dict[str, list[str]] = Field(default_factory=lambda: {
        "trivial": ["ollama", "copilot", "antigravity", "claude"],
        "simple": ["copilot", "antigravity", "codex", "claude"],
        "standard": ["codex", "claude", "antigravity", "copilot"],
        "complex": ["codex", "claude", "antigravity"],
    })
    max_parallel_jobs: int = 3
    max_attempts: int = 3
    job_retention: int = 200


DEFAULT_CLI_AGENTS: list[dict] = [
    {
        "id": "claude", "name": "Claude Code", "supported_tiers": ["simple", "standard", "complex"],
        "quota_reset": "5h", "cost_weight": 1.0, "priority": 40,
    },
    {
        "id": "codex", "name": "OpenAI Codex CLI", "supported_tiers": ["simple", "standard", "complex"],
        "quota_reset": "5h", "cost_weight": 1.0, "priority": 30, "timeout_s": 900,
    },
    {
        "id": "copilot", "name": "GitHub Copilot CLI", "supported_tiers": ["trivial", "simple", "standard"],
        "quota_reset": "weekly", "cost_weight": 0.5, "priority": 50, "max_credits": 10,
    },
    {
        "id": "antigravity", "name": "Google Antigravity CLI (agy)",
        "supported_tiers": ["trivial", "simple", "standard", "complex"],
        "default_model": "gemini-3.1-pro-high",
        "model_by_tier": {"trivial": "gemini-3.8-flash-low", "simple": "gemini-3.8-flash-low",
                          "standard": "gemini-3.1-pro-high", "complex": "gemini-3.1-pro-high"},
        "alt_model_on_quota": "claude-sonnet-4-6",
        "quota_reset": "weekly", "cost_weight": 1.0, "priority": 45,
    },
    {
        "id": "ollama", "name": "Ollama (solo texto)", "supported_tiers": ["trivial"],
        "agentic": False, "cost_weight": 0.0, "priority": 10, "timeout_s": 900,
    },
]


def default_tier_table(provider_ids: set[str]) -> list[TierPolicy]:
    """Tabla inicial que solo referencia providers registrados. Local y barato primero,
    frontier al final; el tier complex prefiere frontier."""
    plan = {
        "trivial": ["ollama", "llamacpp", "lmstudio", "copilot"],
        "simple": ["llamacpp", "ollama", "copilot", "nvidia_nim"],
        "standard": ["copilot", "llamacpp", "deepseek", "anthropic"],
        "complex": ["anthropic", "copilot", "llamacpp"],
    }
    return [
        TierPolicy(tier=tier, targets=[RouteTarget(provider_id=p) for p in ids if p in provider_ids])
        for tier, ids in plan.items()
    ]
